exercises/concept/lasagna/src/test/java/utils/ReflectionProxy.java

```diff
@@ -0,0 +1,477 @@
+     * @param returnType The type of return value
+     * @param name The name of the method
+     * @param parameterTypes The list of method parameters
+     * @return
+     */
+    public boolean isMethodReturnType(Class<?> returnType, String name, Class<?>... parameterTypes) {
+        return isMethodReturnType(returnType, null, name, parameterTypes);
+    }
+
+    /**
+     * Checks if a method from the target class returns a correct parameterized collection (Generics)
+     * @param returnType The return type we expect (eg. List)
+     * @param parameterizedType The parameterized type we expect (eg. String)
+     * @param name The name of the method
+     * @param parameterTypes A list of method parameter types
+     * @return True if the method returns the correct parameterized collection, false otherwise
+     */
+    public boolean isMethodReturnType(Class<?> returnType, Class<?> parameterizedType, String name, Class<?>... parameterTypes) {
+        Class<?> targetClass = getTargetClass();
+        if (targetClass == null || name == null) {
+            return false;
+        }
+        try {
+            Method m = targetClass.getDeclaredMethod(name, parameterTypes);
+            if(!m.getReturnType().equals(returnType)) {
+                return false;
+            }
+            if(parameterizedType == null) {
+                return true;
+            }
+            if(! (m.getGenericReturnType() instanceof ParameterizedType)) {
+                return false;
+            }
+            ParameterizedType pType = (ParameterizedType)m.getGenericReturnType();
+            return pType.getActualTypeArguments()[0].equals(parameterizedType);
+        } catch (NoSuchMethodException e) {
+            return false;
+        }
+    }
+
+    /**
+     * Checks if a method from the target class is public
+     * @param name The name of the method
+     * @param parameterTypes A list of method parameters
+     * @return True if the method exists and is public, false otherwise
+     */
+    public boolean isMethodPublic(String name, Class<?>... parameterTypes) {
+        Class<?> targetClass = getTargetClass();
+        if (targetClass == null || name == null) {
+            return false;
+        }
+        try {
+            Method m = targetClass.getDeclaredMethod(name, parameterTypes);
+            return Modifier.isPublic(m.getModifiers());
+        } catch (NoSuchMethodException e) {
+            return false;
+        }
+    }
+
+    /**
+     * Checks if a target class has a specific constructor
+     * @param parameterTypes The list of desired parameter types
+     * @return True if the constructor exists, false otherwise
+     */
+    public boolean hasConstructor(Class<?>... parameterTypes) {
+        Class<?> targetClass = getTargetClass();
+        if (targetClass == null) {
+            return false;
+        }
+        try {
+            Constructor c = targetClass.getDeclaredConstructor(parameterTypes);
```

Comment:
Constructor<?> c = targetClass.getDeclaredConstructor(parameterTypes);

Wildcard to avoid raw types?

Reply:
done